mini-projects/00-jq-rs/src/filters.rs

```diff
@@ -0,0 +1,156 @@
+use serde_json::Value;
+
+pub fn identity_filter(json: &Value) -> Value {
+    json.clone()
+}
+
+pub fn object_identifier_index(json: &Value, key: &str) -> Value {
+    json.get(key).cloned().unwrap_or(Value::Null)
+}
+
+pub fn array_index(json: &Value, index: usize) -> Value {
+    json.get(index).cloned().unwrap_or(Value::Null)
+}
+
+pub fn array_slice(json: &Value, start: usize, end: usize) -> Value {
+    if let Some(array) = json.as_array() {
+        Value::Array(array[start..end].to_vec())
+    } else {
+        Value::Null
+    }
+}
+
+pub fn pipe(json: &Value, key: &str, index: usize) -> Value {
+    let intermediate = object_identifier_index(json, key);
+    array_index(&intermediate, index)
+}
+
+pub fn array_iterator(json: &Value, key: &str) -> Value {
+    if let Some(array) = json.as_array() {
+        let iterated: Vec<Value> = array
+            .iter()
+            .map(|item| object_identifier_index(item, key))
+            .collect();
+        Value::Array(iterated)
+    } else {
+        Value::Null
+    }
+}
+
+pub fn add(json: &Value) -> Value {
+    if let Some(array) = json.as_array() {
+        let sum = array
+            .iter()
+            .fold(Value::Null, |acc, item| match (acc, item) {
```

Comment:
Good thinking to use functional programming here, but the problem is because the accumulator starts as a `Null`, we can only ever enter the third match case since the other two require a `Number` or `String` accumulator and it never gets set as such.